functorch/csrc/BatchRulesStopDecomposition.cpp

```diff
@@ -6,13 +6,98 @@
 // LICENSE file in the root directory of this source tree.
 
 #include <functorch/csrc/BatchRulesHelper.h>
+#include <ATen/FunctionalTensorWrapper.h>
 #include <ATen/Operators.h>
 #include <functorch/csrc/PlumbingHelper.h>
 #include <functorch/csrc/BatchedFallback.h>
 #include <ATen/core/dispatch/Dispatcher.h>
 
 namespace at { namespace functorch {
 
+at::Tensor sync_and_unwrap_functional_output(at::Tensor out_functional) {
+  TORCH_INTERNAL_ASSERT(at::functionalization::impl::isFunctionalTensor(out_functional));
+  auto out_wrapper_impl = at::functionalization::impl::unsafeGetFunctionalWrapper(out_functional);
+  // sigh.. we can't just call sync(). sync() only runs if the current tensor is "not up to date".
+  // Where "not up to date" is defined as "has any aliases".
+  // Unfortunately at this point, any aliases to `out_functional` will have been deallocated,
+  // So we need to effectively force the sync.
+  out_wrapper_impl->apply_updates();
+  out_wrapper_impl->regenerate_from_base();
+  auto out_unwrapped = out_wrapper_impl->value();
+  return out_unwrapped;
+}
+
+at::TensorList sync_and_unwrap_functional_output(const c10::List<at::Tensor>& t_list) {
+  std::vector<at::Tensor> outputs(t_list.size());
+  for (const auto i : c10::irange(t_list.size())) {
+    outputs[i] = sync_and_unwrap_functional_output(t_list[i]);
+  }
+  return outputs;
+}
+
+void decompose_functional(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+  const auto& schema = op.schema();
+
+  const auto num_arguments = schema.arguments().size();
+  const auto arguments = torch::jit::last(stack, num_arguments);
+  const auto arguments_begin = stack->size() - num_arguments;
+  //
+  // Step 1: Wrap any tensor inputs into Functional tensors
+  // and put them on the stack at the correct indices.
+  for (const auto idx : c10::irange(arguments.size())) {
+    const auto& ivalue = arguments[idx];
+    if (ivalue.isTensor()) {
+      auto functional_ivalue = at::functionalization::impl::to_functional_tensor(ivalue.toTensor());
+      (*stack)[arguments_begin + idx] = std::move(functional_ivalue);
+    } else if (ivalue.isTensorList()) {
+      auto functional_ivalue = at::functionalization::impl::to_functional_tensor(ivalue.toTensorList());
+      (*stack)[arguments_begin + idx] = std::move(functional_ivalue);
+    }
+  }
+
+  // Step 2: set up TLS such that we hit the functionalization kernels before the batching rules.
+  // Note: this relies on the fact that Functionalization > BatchMode in DispatchKey.h
```

Comment:
I think I understand the problem you were mentioning last week where it is unclear how to "push the functionalize transform onto the dynamic layer stack" -- when we get to DECOMPOSE_FUNCTIONAL, the dynamic layer stack has already decided to dispatch to the vmap subsystem!